content/6_Advanced/Critical.mdx

```diff
@@ -10,17 +10,318 @@ frequency: 1
+4. Let $w \neq S$. Let u be a vertex for which $sdom(u)$ is minimum among all vertices u satisfying “$sdom(w)$ is a proper ancestor of $u$ and $u$ is an ancestor of $w$”.  If $sdom(u) \leq sdom(w)$ then $idom(w) = idom(u)$.
+5. Let $u \neq S$ and &v& be a vertex for which $sdom(v)$ is minimum among all vertices $u$ satisfying: $sdom(u)$ is a proper ancestor of $v$ and $v$ is an ancestor of u.  If $sdom(v) \le sdom(u)$ then $idom(u) = idom(v)$.
+6. Let $w \neq S$. Let $u$ be a vertex for which $sdom(u)$ is minimum among all vertices $u$ satisfying “$sdom(w)$ is a proper ancestor of $u$ and $u$ is an ancestor of $w$”. Then:
+
+$$
+idom(w)=\begin{cases}
+sdom(w), & \text{ If } sdom(u)=sdom(w),\\
+idom(w), & \text{otherwise}.
+\end{cases}
+\tag{1}
+$$
+
+## Precompute Sdom
+
+The following identity enables us the compute $sdom$:
+$$
+
+sdom(u) = min((y \vert (y, u) \in E \text{ and } y \le u) \cup (sdom(x) \vert x>u \\\text{ and there is and edge (x, y) such that x is ancestor of y}) )
+\tag{2}
+$$
+In other words, $sdom(u)$ is the minimum node in the intersection of the following groups:
+1. All the nodes $y$ with $y \le u \text{ and } (y,u) \in E$
+2. $sdom(x)$ where $x > u \text{ and } (u, y) \in E$
+
+## Implementation
+
+1. Precompute using a DFS the new mapping of the tree, i.e. the new labels of the nodes are their entry time $u = e[u]$.
+
+2. Compute $sdom$ for all nodes by applying the formula mentioned in the previous section. Iterate over node in decreasing order, maintaining the visited nodes in a DSU. Also, the DSU is slightly modified:
+
+- $Find(u)$ = Let $x$ be the root of the components
+    - If $u=x$, return $x$;
+    - else return a node $v$ with minimum sdom, lying on path from $u$ to $x$.
+
+In order to process node $u$ we iterate over all incoming edges $(v, u):$
+- If $v < u$, $v$ is an ancestor of $u$, $v$ would not have been processed till now, thus $Find(v)$ would return v.
+- If $v > w$ then $v$ would have already been processed and $Find(v)$ would return a node $x$ lying on the path from $v$ to root in the DSU tree having minimum sdom.
+
+The 2 cases cover precisely the definition $(2)$.
+
+3. Compute the $idom$ for all nodes by applying $(1)$.
+
+In this step, store nodes in buckets of their own $sdom$ and iterate over them updating the $idom$ value using $(1)$.
+
+This snippet of code combines steps $2$ and $3$:
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+// Iterate decreasingly
+for (int i = n - 1; i >= 0; i--) {
+	// Iterat over incoming edges applying (2)
+	for (int p : rg[i]) { sdom[i] = min(sdom[i], sdom[Find(p, 0)]); }
+	// Insert i into it's bucket after sdom[i]
+	if (i > 0) { bucket[sdom[i]].push_back(i); }
+	// Apply (1) for nodes in bucket
+	for (int node : bucket[i]) {
+		int p = Find(node, 0);
+		if (sdom[p] == sdom[node]) {
+			dom[node] = sdom[node];
```

Comment:
space chunks of code out a little bit

Reply:
ok

Reply:
not fixed?

Reply:
dw ryan i'm boutta sansify this whole module